plugins/net_plugin/net_plugin.cpp

```diff
@@ -4486,39 +4490,48 @@ namespace eosio {
       return resolve_and_connect( host, p2p_address );
    }
 
-   string connections_manager::resolve_and_connect( const string& peer_address, const string& listen_address ) {
+   string connections_manager::resolve_and_connect( const string& peer_address, const string& listen_address,
+                                                    const connection_ptr& c )
+   {
+      assert(!c || (c->peer_address() == peer_address && c->listen_address == listen_address));
+
       string::size_type colon = peer_address.find(':');
       if (colon == std::string::npos || colon == 0) {
          fc_elog( logger, "Invalid peer address. must be \"host:port[:<blk>|<trx>]\": ${p}", ("p", peer_address) );
          return "invalid peer address";
       }
 
-      std::lock_guard g( connections_mtx );
-      if( find_connection_i( peer_address ) )
-         return "already connected";
+      {
+         std::lock_guard g(connections_mtx);
+         if (find_connection_i(peer_address))
+            return "already connected";
+      }
 
       auto [host, port, type] = split_host_port_type(peer_address);
 
       auto resolver = std::make_shared<tcp::resolver>( my_impl->thread_pool.get_executor() );
 
       resolver->async_resolve(host, port, 
-         [resolver, host = host, port = port, peer_address = peer_address, listen_address = listen_address, this]( const boost::system::error_code& err, const tcp::resolver::results_type& results ) {
-            connection_ptr c = std::make_shared<connection>( peer_address, listen_address );
-            c->set_heartbeat_timeout( heartbeat_timeout );
-            std::lock_guard g( connections_mtx );
-            auto [it, inserted] = connections.emplace( connection_detail{
-               .host = peer_address,
-               .c = std::move(c),
-               .ips = results
+         [this, resolver, c_org{c}, host, port, peer_address, listen_address]( const boost::system::error_code& err, const tcp::resolver::results_type& results ) {
+            connection_ptr c = c_org ? c_org : std::make_shared<connection>( peer_address, listen_address );
+            c->strand.post([this, resolver, c, err, results, host, port, peer_address]() {
+               c->set_heartbeat_timeout( heartbeat_timeout );
+               {
+                  std::lock_guard g( connections_mtx );
+                  connections.emplace( connection_detail{
+                     .host = peer_address,
+                     .c = c,
+                  });
+               }
+               if( !err ) {
+                  c->connect( results );
+               } else {
+                  fc_wlog( logger, "Unable to resolve ${host}:${port} ${error}",
```

Comment:
fwiw this is a warning log, where failing to connect is an info log. Not sure if you want to make consistent or not

Reply:
If the `async_connect` fails it calls `c->close(false)` which calls `_close()` which increments `consecutive_immediate_connection_close`.

Reply:
I think a `warn` is appropriate if unable to resolve.

Reply:
> If the `async_connect` fails it calls `c->close(false)` which calls `_close()` which increments `consecutive_immediate_connection_close`.

There is something different between the two though. For example when I run nodeos with ` --p2p-peer-address fdsfdsfds.invalid:4444 --p2p-peer-address fddsfdsfdsfds.invalid:4444 --p2p-peer-address 127.0.0.1:2121 --p2p-peer-address www.google.com:4444` you'll see how the latter 2 will try to reconnect forever where the first 2 won't.

There is also some log oddness  at shutdown,
```
info  2024-09-26T02:45:43.993 nodeos    net_plugin.cpp:4578           close_all            ] close all 4 connections
info  2024-09-26T02:45:43.993 net-1     net_plugin.cpp:1460           _close               ] ["127.0.0.1:2121" - 1 <unknown>:<unknown>] closing
info  2024-09-26T02:45:43.993 net-2     net_plugin.cpp:1460           _close               ] ["fddsfdsfdsfds.invalid:4444" - 2 <unknown>:<unknown>] closing
info  2024-09-26T02:45:43.993 nodeos    net_plugin.cpp:4385           plugin_shutdown      ] exit shutdown
```

Reply:
This different in backoff behavior seems to match Leap 4.0 behavior.